### DISCUSSION TITLE
How to add new JSBsim component in unreal engine 5?

### DISCUSSION THREAD

**Discussion by cxr001** [JSBSim-Team/jsbsim#812] [Support requests]
I want to set vUVW value after running the engine, but I don't know how to write these code in JSBsim, and transfer them into UE5.
I also not sure how to use the autopilot in UE.

**Comment by cxr001** [comment]
Is there any more detailed tutorial for the JSBsim UE project? I am new to JSBsim and UE, and it is a little hard to learn all of these things.

**Comment by seanmcleod** [comment]
> I want to set vUVW value after running the engine

Do you want to set `UVW` just once during initialization or are you wanting to change it while the simulation is running?

`UVW` can't be changed during simulation, they're read-only.

```c++
  PropertyManager->Tie("velocities/u-aero-fps", this, eU, (PMF)&FGAuxiliary::GetAeroUVW);
  PropertyManager->Tie("velocities/v-aero-fps", this, eV, (PMF)&FGAuxiliary::GetAeroUVW);
  PropertyManager->Tie("velocities/w-aero-fps", this, eW, (PMF)&FGAuxiliary::GetAeroUVW);
```

But you can set `UVW` as part of the initial conditions during initialization.

**Reply by cxr001** [reply]
I wish the airplane can realize different take-off methods, for example, by ejection. So the airplane should have a zero speed at first, although it has started the engine and has had some RPM at that time. Then start the engine, the airplane would have a speed suddenly, and these speed will be used just once. Could this be realised?

**Comment by seanmcleod** [comment]
When you say ejection, which normally refers to a pilot ejecting from an aircraft, do you mean you want some sort of vertical take-off?

Give us some more information on what you're trying to model and then we'll have a better chance of suggesting useful solutions.

**Reply by cxr001** [reply]
I mean a type of take-off, it will have a 0 speed but have started the engine at first, and then it would be given a force along the nose

**Comment by seanmcleod** [comment]
So use the initial conditions to set the engine running and then set the throttle to it's maximum.

**Reply by cxr001** [reply]
That speed is not enough. and the running engine would let the aircraft have a speed. I find that there is a file named FGExternalForce, could I use that in UE?

**Comment by seanmcleod** [comment]
Are you trying to model something like a catapult launch from an aircraft carrier? Where the engine is run up to maximum power and then the catapult provides an additional force for a particular distance? More details will help.

**Reply by cxr001** [reply]
Yes, it is something like a catapult launch

**Comment by seanmcleod** [comment]
Yep then take a look at using an external force.

**Reply by cxr001** [reply]
But how to set the speed as zero when the engine is running?

**Reply by bcoconni** [reply]
You could set the property `forces/hold-down` to 1.0 to avoid aircraft movement while you are initializing the aircraft settings.

This feature is used for instance to simulate rocket engines ignition and thrust build up before releasing the rocket (by setting `forces/hold-down` to 0.0).
https://github.com/JSBSim-Team/jsbsim/blob/97f7b20421eaa8a0587e3a6f0e43645cae5a6c62/scripts/J2461.xml#L96-L100

**Reply by cxr001** [reply]
Thank you, I get it. and how to initialize the aircraft velocities or forces in UE?

**Reply by bcoconni** [reply]
> Thank you, I get it. and how to initialize the aircraft velocities or forces in UE?

@gallonmate any idea ?

**Reply by gallonmate** [reply]
@cxr001 would need to confirm what they are asking. I believe they are asking which property can be used to apply some External Force on an aircraft, to simulate a catapult or external force launching a vehicle. But I do not know External Force so I cannot help there.
@cxr001 now appears to successfully be setting properties in Unreal Engine, seen in #813. So they would only need to know the property name and value to be using.

**Reply by cxr001** [reply]
Yes, I need to know the name of property, which can be used to apply some External Force, or any property to set the initial conditions of the UAV's speed ,or let the engine be idling.

**Reply by bcoconni** [reply]
There are quite a number of examples in the aircraft models that are provided with JSBSim. Just search for `<external_reactions>` in the `aircraft` folder.

Below are a few examples (you'll need to scroll on some of them to review the whole code snippet) :
https://github.com/JSBSim-Team/jsbsim/blob/9a99b4b60b3ce00c4308c3a2539a7712824e2f17/aircraft/787-8/787-8.xml#L980-L993
https://github.com/JSBSim-Team/jsbsim/blob/9a99b4b60b3ce00c4308c3a2539a7712824e2f17/aircraft/ball/ball.xml#L100-L132
https://github.com/JSBSim-Team/jsbsim/blob/9a99b4b60b3ce00c4308c3a2539a7712824e2f17/aircraft/Boeing314/Boeing314.xml#L1630-L1646
https://github.com/JSBSim-Team/jsbsim/blob/9a99b4b60b3ce00c4308c3a2539a7712824e2f17/aircraft/Concorde/Concorde.xml#L1734-L1747

**Comment by seanmcleod** [comment]
You can have the engine running but with a zero value for the throttle, i.e. the engine just idling which would prevent the aircraft from starting to move until you increase the throttle and/or apply the external force from the catapult. 

Or if you want to have the engine at full thrust then you'll need to apply the brakes to prevent any movement until you release the brakes.

**Reply by cxr001** [reply]
which function could I use to let the engine be idling?
